File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        cur=root
        def height(cur):
                if not cur:
                    return 0
                return 1 + max(height(cur.left),height(cur.right))
        def check(cur):
            if not cur:
                return True
            l=height(cur.left)
            r=height(cur.right)
            if abs(l-r)>1:
                return False
            lh=check(cur.left)
            rh=check(cur.right)
            # if (not lh or not rh):
            #     return False
            # return True
            return lh and rh
        return check(cur)

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        
        int left = maxDepth(root.left);
        int right = maxDepth(root.right);
        boolean check = false;
        if (Math.abs(left - right) <= 1) check = true;
        return (check && isBalanced(root.left) && isBalanced(root.right));
    }
    
    private int maxDepth(TreeNode node){
        if (node == null) return 0;
        int left = 1 + maxDepth(node.left);
        int right = 1 + maxDepth(node.right);
        return Math.max(left, right);
    }
}